src/diffusers/models/controlnet_xs.py

```diff
@@ -0,0 +1,1228 @@
+            interpolation_param = self.control_addon.config.time_embedding_mix**0.3
+
+            temb = ctrl_temb * interpolation_param + base_temb * (1 - interpolation_param)
+        else:
+            temb = self.base_time_embedding(t_emb)
+
+        # added time & text embeddings
+        aug_emb = None
+
+        if self.base_class_embedding is not None:
+            if class_labels is None:
+                raise ValueError("class_labels should be provided when num_class_embeds > 0")
+
+            if self.config.class_embed_type == "timestep":
+                class_labels = self.base_time_proj(class_labels)
+
+            class_emb = self.base_class_embedding(class_labels).to(dtype=self.dtype)
+            temb = temb + class_emb
+
+        if self.config.addition_embed_type is None:
+            pass
+        elif self.config.addition_embed_type == "text_time":
+            # SDXL - style
+            if "text_embeds" not in added_cond_kwargs:
+                raise ValueError(
+                    f"{self.__class__} has the config param `addition_embed_type` set to 'text_time' which requires the keyword argument `text_embeds` to be passed in `added_cond_kwargs`"
+                )
+            text_embeds = added_cond_kwargs.get("text_embeds")
+            if "time_ids" not in added_cond_kwargs:
+                raise ValueError(
+                    f"{self.__class__} has the config param `addition_embed_type` set to 'text_time' which requires the keyword argument `time_ids` to be passed in `added_cond_kwargs`"
+                )
+            time_ids = added_cond_kwargs.get("time_ids")
+            time_embeds = self.base_add_time_proj(time_ids.flatten())
+            time_embeds = time_embeds.reshape((text_embeds.shape[0], -1))
+            add_embeds = torch.concat([text_embeds, time_embeds], dim=-1)
+            add_embeds = add_embeds.to(temb.dtype)
+            aug_emb = self.base_add_embedding(add_embeds)
+        else:
+            raise ValueError(
+                f"ControlNet-XS currently only supports StableDiffusion and StableDiffusion-XL, so addition_embed_type = {self.config.addition_embed_type} is currently not supported."
+            )
+
+        temb = temb + aug_emb if aug_emb is not None else temb
+
+        # text embeddings
+        cemb = encoder_hidden_states
+
+        # Preparation
+        h_ctrl = h_base = sample
+        hs_base, hs_ctrl = [], []
+
+        # Cross Control
+        # Let's first define variables to shorten notation
+        base_down_subblocks = self.base_down_subblocks
+        ctrl_down_subblocks = self.control_addon.down_subblocks
+
+        down_zero_convs_b2c = self.control_addon.down_zero_convs_b2c
+        down_zero_convs_c2b = self.control_addon.down_zero_convs_c2b
+        mid_zero_convs_c2b = self.control_addon.mid_zero_convs_c2b
+        up_zero_convs_c2b = self.control_addon.up_zero_convs_c2b
+
+        if not do_control:
```

Comment:
instead of these sub_blocks, can we create `ControlnetXsCrossAttnDownBlock2D`, `ControlNetXSDownBlock2D` that:
1. behave exactly the same as `CrossAttnDownBlock2D` and `DownBlock2D` when `do_control = False`
2. do whatever these subblock are doing when `do_control = True`

Reply:
the idea is that instead of having a `base_model` and a control_addon` and using their sub-blocks in the forward pass, we will try to make a `UNetControlNetXSModel` that follows the general structure of `UNet2DConditionModel`, e.g. it will also have`down_blocks`, which consists of `ControlnetXsCrossAttnDownBlock2D` and `ControlNetXSDownBlock2D` that I mentioned above 

same goes for `up_blocks`

Reply:
@yiyixuxu Done! Happy for a review :)